File: JavaScript/Medium/1d_programming/coin_change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */

const char = 1;

var coinChnage = function (coins, amount) {
    if (amount == 0){ return 0}
    coins = coins.sort((a,b)=>b-a)

    let minCount = new Array(coins.length+1).fill(Infinity)
    minCount[0]=0

    for (let num = 1; num<=amount; num++){
        for (let coin of coins){
            if (coin > num) { break }

            minCount[num] = Math.min(minCount[num], char+minCount[num-coin])
        }
    }
}

coins = [3,1,5]
amount=3
coinChnage(coins, amount)

File: Python/Medium/1d-programming/coin_change.py
class Solution(object):
    def coinChange(coins, amount):
        if amount == 0: return 0
        coins.sort()

        min_count = [float('inf')]*(amount+1)
        min_count[0]=0

        for i in range(1,amount+1):
            for coin in coins:
                if coin>i:
                    break
                min_count[i]=min(min_count[i], 1+min_count[i-coin])

        if min_count[amount] == float('inf'):
            return -1
        else:
            return min_count[amount]
        